content/blog/testthat-3-3-0/index.Rmd

```diff
@@ -0,0 +1,172 @@
+
+* All expectations have new failure messages: they now state what was expected, what was actually received, and, if possible, they clearly illustrate the difference.
+
+* Expectations now consistently return the value of the first argument, regardless of whether the expectation succeeds or fails (the only exception is `expect_error()` and friends which return the captured condition). This is a relatively subtle change that won't affect tests that already pass, but it does improve failures when you pipe together multiple expectations.
+
+* A new `pass()` function makes it clear how to signal when an expectation succeeds. All existing expectations were rewritten to use `pass()` and (the existing) `fail()` instead of `expect()`, which I think makes the flow of logic easier to understand.
+
+* Improved `expect_success()` and `expect_failure()` expectations now test that an expectation always returns exactly one success or failure (this ensures that the counts that you see in the reporters are correct).
+
+This new framework helped us write six new expectations:
+
+*   `expect_all_equal()`, `expect_all_true()`, and `expect_all_false()` check that every element of a vector has the same value, giving better error messages than `expect_true(all(...))`:
+
+    ```{r}
+    #| error: true
+
+    test_that("some test", {
+      x <- c(0.408, 0.961, 0.883, 0.46, 0.537, 0.961, 0.851, 0.887, 0.023)
+      expect_all_true(x < 0.95)
+    })
+    ```
+
+*   `expect_disjoint()`, by [@stibu81](https://github.com/stibu81), expects values to be absent:
+
+    ```{r}
+    #| error: true
+
+    test_that("", {
+      expect_disjoint(c("a", "b", "c"), c("c", "d", "e"))
+    })
+    ```
+
+*   `expect_r6_class()` expects an R6 object:
+
+    ```{r}
+    #| error: true
+
+    test_that("", {
+      x <- 10
+      expect_r6_class(x, "foo")
+
+      x <- R6::R6Class("bar")$new()
+      expect_r6_class(x, "foo")
+    })
+    ```
+
+*   `expect_shape()`, by [@michaelchirico](https://github.com/michaelchirico), expects a specific shape (i.e., `nrow()`, `ncol()`, or `dim()`):
+
+    ```{r}
+    #| error: true
+
+    test_that("show off expect_shape() failure messages", {
+      x <- matrix(1:9, nrow = 3)
+      expect_shape(x, nrow = 4)
+      expect_shape(x, dim = c(3, 3, 3))
+      expect_shape(x, dim = c(3, 4))
+    })
+    ```
+
+As you can see from the examples above, when you run a single test interactively (i.e. not as a part of a test suite) you now see exactly how many expectations succeeded and failed.
+
+## Other new features
+
+* testthat generally does a better job of handling nested tests, aka subtests, where you put a `test_that()` inside another `test_that()`, or more typically `it()` inside of `describe()`. Subtests will now generate more informative failure messages, free from duplication, with more informative skips if any subtests don't contain any expectations.
```

Comment:
I thought nesting `test_that()` was mostly useful for testing the testthat package with itself, but this paragraph makes it sound like there might be user-oriented cases where this is helpful? If that's the case, it might be interesting to add a sentence explaining the use case, otherwise a sentence explaining this is mostly for internal testing.